db/index.go

```diff
@@ -0,0 +1,495 @@
+		}
+	}()
+
+	for res := range q.Next() {
+		if res.Error != nil {
+			return res.Error
+		}
+		err = txn.Datastore().Delete(ctx, ds.NewKey(res.Key))
+		if err != nil {
+			return NewCanNotDeleteIndexedField(err)
+		}
+	}
+
+	return nil
+}
+
+func (i *collectionSimpleIndex) Name() string {
+	return i.desc.Name
+}
+
+func (i *collectionSimpleIndex) Description() client.IndexDescription {
+	return i.desc
+}
+
+func validateIndexDescription(desc client.IndexDescription) error {
+	if desc.ID != 0 {
+		return NewErrNonZeroIndexIDProvided(desc.ID)
+	}
+	if len(desc.Fields) == 0 {
+		return ErrIndexMissingFields
+	}
+	if len(desc.Fields) == 1 && desc.Fields[0].Direction == client.Descending {
+		return ErrIndexSingleFieldWrongDirection
+	}
+	for i := range desc.Fields {
+		if desc.Fields[i].Name == "" {
+			return ErrIndexFieldMissingName
+		}
+		if desc.Fields[i].Direction == "" {
+			desc.Fields[i].Direction = client.Ascending
+		}
+	}
+	return nil
+}
+
+func generateIndexName(col client.Collection, fields []client.IndexedFieldDescription, inc int) string {
+	sb := strings.Builder{}
+	direction := "ASC"
+	//if fields[0].Direction == client.Descending {
+	//direction = "DESC"
+	//}
+	sb.WriteString(col.Name())
+	sb.WriteByte('_')
+	sb.WriteString(fields[0].Name)
+	sb.WriteByte('_')
+	sb.WriteString(direction)
+	if inc > 1 {
+		sb.WriteByte('_')
+		sb.WriteString(strconv.Itoa(inc))
+	}
+	return sb.String()
+}
+
+func (c *collection) CreateIndex(
+	ctx context.Context,
+	desc client.IndexDescription,
+) (client.IndexDescription, error) {
+	index, err := c.createIndex(ctx, desc)
+	if err != nil {
+		return client.IndexDescription{}, err
+	}
+	if c.isIndexCached {
+		c.indexes = append(c.indexes, index)
+	}
+	return index.Description(), nil
+}
+
+func (c *collection) DropIndex(ctx context.Context, indexName string) error {
+	key := core.NewCollectionIndexKey(c.Name(), indexName)
+
+	txn, err := c.getTxn(ctx, false)
+	if err != nil {
+		return err
+	}
+	_, err = c.getIndexes(ctx, txn)
+	if err != nil {
+		return err
+	}
+	for i := range c.indexes {
+		if c.indexes[i].Name() == indexName {
+			err = c.indexes[i].RemoveAll(ctx, txn)
+			if err != nil {
+				return err
+			}
+			c.indexes = append(c.indexes[:i], c.indexes[i+1:]...)
+			break
+		}
+	}
+	err = txn.Systemstore().Delete(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *collection) dropAllIndexes(ctx context.Context) error {
+	prefix := core.NewCollectionIndexKey(c.Name(), "")
+	txn, err := c.getTxn(ctx, false)
+	if err != nil {
+		return err
+	}
+	q, err := txn.Systemstore().Query(ctx, query.Query{
+		Prefix: prefix.ToString(),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := q.Close(); err != nil {
+			log.ErrorE(ctx, "Failed to close collection query", err)
+		}
+	}()
+
+	for res := range q.Next() {
+		if res.Error != nil {
+			return res.Error
+		}
+		err = txn.Systemstore().Delete(ctx, ds.NewKey(res.Key))
```

Comment:
question: How sure are you that this works?  It is unusual for mutation of an enumerable that you are currently iterating through to be safe/correct.

Reply:
> But now thinking about this I realized that Badger is not the only db that can be used with defra. So it's better to separate those actions to be on the safe side.

That's a good point, happy with as is if you can find similar documentation on the ipfs interface declarations though.